core interrupt: use device tree data

etienne-lms opened the pull request:
Following changes done recently in the interrupt framework, this P-R is based on https://github.com/OP-TEE/optee_os/pull/6097. It provides means to use DT bindings to describe interrupts and their controllers.

2nd commit adds an API function (`dt_register_interrupt_provider()`)for interrupt controller drivers to register as itr_chip providers into the DT_DRIVER framework and few API functions for consumer driver to use DT info to find their interrupt (`interrupt_dt_get()`, `interrupt_dt_get_by_index()`, `interrupt_dt_get_by_name()`).

3rd commit Adds a API function to register a callback handler for an interrupt. 
`interrupt_create_handler()` differs from existing `interrupt_add_handler()` in that the former does not reconfigure the interrupt since already done by `interrupt_dt_get()` or friend. Maybe `interrupt_add_handler2()` would be a better label.

Last commit registers GIC driver as a `DT_DRIVER_INTERRUPT` driver so that interrupt consumer drivers can use `interrupt_dt_get_by_index()` and friends.

etienne-lms commented on core/drivers/gic.c:
```diff
@@ -604,3 +605,67 @@ static void gic_op_set_affinity(struct itr_chip *chip, size_t it,
+	 * formats. Therefore swap again phandle arguments.
+	 */
+	for (n = 0; n < arg->args_count; n++)
+		arg->args[n] = cpu_to_fdt32(arg->args[n]);
```
This is not nice. It modifies `arg` content. If called several times (some late init deferral), them `arg` content will not be reliable. I'll change into using a local copy of the arguments on which to call `cpu_to_fdt32()`.

etienne-lms commented:
I've squashed the fixup commits.

jforissier commented on core/include/kernel/interrupt.h:
```diff
@@ -343,4 +358,113 @@ TEE_Result interrupt_alloc_add_handler(struct itr_chip *chip, size_t it_num,
+ * Return another TEE_Result code otherwise.
+ *
+ * Upon success, struct itr_desc must point to memory allocated with malloc()
+ * or like as it is freed prior returning to caller function.
```
"freed with free()"?

jforissier reviewed (commented):
One small comment for "core: interrupt: registering interrupt providers", with or without it:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
"core: dt_driver: don't enforce phandle 1st arg is a phandle":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
"core: interrupt: add interrupt_create_handler()":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
"drivers: gic: register to dt_driver":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
"drivers: gic: use DT bindings":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
https://github.com/OP-TEE/optee_os/pull/6361#discussion_r1368734996 addressed.
Tags applied.
Thanks for the review.

I've created a P-R in optee_docs for these new API functions: https://github.com/OP-TEE/optee_docs/pull/218.

jforissier commented:
Please rebase onto master to fix the IBART error, thanks!

etienne-lms commented:
right, thanks.

etienne-lms commented:
rebased

etienne-lms commented:
Arrg! IBART still fails.

etienne-lms commented:
I did not expect this change would make such platforms to consume more heap. IBART does not use DT for OP-TEE drivers inits.

jenswi-linaro commented on core/kernel/interrupt.c:
```diff
@@ -196,3 +196,129 @@ void interrupt_remove_free_handler(struct itr_handler *hdl)
+		*itr_num = desc->itr_num;
+
+		/* Balance malloc() or like from itr_dt_get_func callback */
+		free(desc);
```
This is odd. Why not allocate this on the stack?

etienne-lms replied:
Makes sense. I think it's feasible.
`get_of_device_func` function type and `dt_driver_device_from_xxx()` currently expect the reference is an output only ref, but I think it's fine if also used as a input argument. I'll propose something.

etienne-lms commented:
Comment addressed.

jenswi-linaro commented on core/drivers/gic.c:
```diff
@@ -610,11 +610,11 @@ static void gic_op_set_affinity(struct itr_chip *chip, size_t it,
 /* Callback for "interrupts" and "interrupts-extended" DT node properties */
 static TEE_Result dt_get_gic_chip_cb(struct dt_pargs *arg, void *priv_data,
-				     struct itr_desc **out_itr_desc)
+				     struct itr_desc **itr_desc_p)
```
Why is this a double-pointer? I can't see that it's needed anywhere and it makes the code more complicated.

etienne-lms replied:
The generic `dt_driver_device_from_xxx()` functions expect a `void **` reference (e.g. [`dt_driver_device_from_node_idx_prop()`](https://github.com/OP-TEE/optee_os/blob/4.0.0/core/include/kernel/dt_driver.h#L156)), hence i used a double pointer reference here also, for consistency. Maybe over killing.

jenswi-linaro replied:
I don't quite see what's consistent here, but it does look strange and error-prone to me.

etienne-lms replied:
Ok, i'll add a fixup patch in the hope I understood your concern.

etienne-lms commented:
rebased.

jenswi-linaro commented:
If "core: kernel: dt_driver: reference output device reference as void *" was possible before, why did we ever have pointer-pointers as arguments to those functions?

etienne-lms commented:
My bad. The several driver frameworks use `struct clk **`, `struct rstctrl **`, `struct regulator **`, etc... hence I chose `void **` generic type in _core/kernel/dt_driver.c_ to represent that (see commit https://github.com/OP-TEE/optee_os/commit/b357d34fe91f4e7f6e0eacea17a7fbe5f6c01e7e) in the generic functions the several device frameworks rely on. Maybe a bad idea, I could have used `void *` at the time.

Note I will place commit "core: kernel: dt_driver: reference output device reference as void *" at the beginning of the series when I will squash the fixup commits.

etienne-lms commented:
I've squashed the fixup commits and reordered the series.

@jforissier, I removed your A-b tag from commits "core: interrupt: registering interrupt providers"
and "drivers: gic: register to dt_driver" as content sligthly  changed since you reviewed them.

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
> I've squashed the fixup commits and reordered the series.
> 
> @jforissier, I removed your A-b tag from commits "core: interrupt: registering interrupt providers" and "drivers: gic: register to dt_driver" as content sligthly changed since you reviewed them.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>` for both, thanks.

etienne-lms commented:
@jbech-linaro, could you help me understand why IBART fails with this series? https://github.com/OP-TEE/optee_os/pull/6361#issuecomment-1787195610 mentions a core heap exhaustion but since the platform used by IBART does not embed any secure DTB, I don't understand why these changes make the platform to more core heap.

etienne-lms commented:
I forgot to mention that i've applied the review tags.

jbech-linaro commented:
As discussed in email. I've been able to reproduce the error on two difference machines. So, although I don't now that causes it, it at least seems reproducible.

etienne-lms commented:
IBART CI test failure should be fixed by https://github.com/OP-TEE/optee_test/pull/711.

etienne-lms commented:
IBART's happy!
